### PR TITLE
fix(docker): avoid setup CLI namespace loop before gateway startup

### PR DESCRIPTION
**Pull request by amsminn**
## Summary

- Problem: `docker-setup.sh` runs several `docker compose run openclaw-cli ...` commands before the gateway is started, but `openclaw-cli` now shares the gateway network namespace.
- Why it matters: fresh Docker installs can deadlock into a crash-loop where the gateway needs `allowedOrigins` before startup, while the setup script needs a working gateway namespace before it can write that config.
- What changed: pre-start setup commands now run as one-shot `openclaw-gateway` containers with `--no-deps` and `--entrypoint node`, so config writes and onboarding no longer depend on a running gateway; the default seeded allowlist now includes both `http://localhost:<port>` and `http://127.0.0.1:<port>`; the Docker setup e2e test now locks in that no shared-network `openclaw-cli` run happens before `openclaw-gateway` is started.
- What did NOT change (scope boundary): no runtime gateway auth/origin enforcement changed, and no sandbox/runtime Docker wiring changed after the gateway is already up.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [x] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [x] CI/CD / infra

## Linked Issue/PR

- Related #39228
- Related #41114
- Related #27473
- Related #48412
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: `docker-compose.yml` from #12504 moved `openclaw-cli` to `network_mode: "service:openclaw-gateway"`, but `scripts/docker/setup.sh` still used `docker compose run --rm openclaw-cli ...` for pre-start setup writes.
- Missing detection / guardrail: there was no test asserting that setup-stage CLI invocations remain independent of the gateway namespace before the gateway has started.
- Prior context (`git blame`, prior PR, issue, or refactor if known): this regressed after #12504 introduced the shared network namespace to fix post-start CLI connectivity.
- Why this regressed now: non-loopback binds require `gateway.controlUi.allowedOrigins` before startup, so the setup script hit a circular dependency introduced by the new shared namespace.
- If unknown, what was ruled out: N/A.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [ ] Unit test
  - [ ] Seam / integration test
  - [x] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `src/docker-setup.e2e.test.ts`
- Scenario the test should lock in: before `docker compose up -d openclaw-gateway`, `docker-setup.sh` must not run the shared-network `openclaw-cli` service.
- Why this is the smallest reliable guardrail: the failure depends on Compose service wiring plus setup-script ordering, not pure function behavior.
- Existing test that already covers this (if any): the existing Docker setup e2e test already covers the script flow and was extended here.
- If no new test is added, why not: N/A.

## User-visible / Behavior Changes

Fresh Docker setup no longer crash-loops when pre-start config writes need to run before the gateway exists. When the script seeds default Control UI origins for a non-loopback bind, it now writes both `localhost` and `127.0.0.1` loopback origins.

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: macOS 15.4 host for validation; repro also confirmed from latest `origin/main` in an isolated clone via a Docker stub that fails any pre-start shared-network `openclaw-cli` run.
- Runtime/container: Docker Compose path in `scripts/docker/setup.sh`
- Model/provider: N/A
- Integration/channel (if any): N/A
- Relevant config (redacted): `OPENCLAW_GATEWAY_BIND=lan`, `OPENCLAW_GATEWAY_PORT=18789`

### Steps

1. Clone latest `origin/main` and run `docker-setup.sh` with non-loopback Docker defaults.
2. Observe that setup runs `docker compose run ... openclaw-cli ...` before the gateway is stably up.
3. Gateway restart loop prevents the CLI sidecar from joining its network namespace, so setup aborts.

### Expected

- Setup can complete its config writes before the gateway is started.
- No `cannot join network namespace ... is restarting` error during the `Docker gateway defaults` / allowlist stage.

### Actual

- On latest `origin/main`, a stubbed repro exits with `Error response from daemon: cannot join network namespace of container ... is restarting, wait until the container is running`.

## Evidence

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)

- Verified scenarios: rebased the fix onto latest `origin/main`, installed dependencies, ran `corepack pnpm test -- src/docker-setup.e2e.test.ts`, and manually reproduced the pre-fix failure against latest `origin/main` with a Docker stub that rejects pre-start `openclaw-cli` runs.
- Edge cases checked: rerun guard still reads existing `gateway.controlUi.allowedOrigins` before writing; default seeding now covers both `localhost` and `127.0.0.1`.
- What you did **not** verify: live end-to-end Docker run against a real daemon after rebase; sandbox follow-up paths beyond the existing script behavior.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? No new knobs; the setup script seeds a slightly broader default allowlist when it is responsible for creating that value.
- Migration needed? No
- If yes, exact upgrade steps: N/A

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: revert the two commits in this PR.
- Files/config to restore: `scripts/docker/setup.sh`, `src/docker-setup.e2e.test.ts`
- Known bad symptoms reviewers should watch for: setup unexpectedly opening gateway service ports during one-shot runs; onboarding/config writes failing because the one-shot gateway container no longer behaves like the CLI image path.

## Risks and Mitigations

- Risk: using the `openclaw-gateway` service as the one-shot setup runner could drift from `openclaw-cli` if the service environments diverge later.
  - Mitigation: both services currently use the same image, config/workspace mounts, token env, and HOME; the new e2e assertion locks in the pre-start dependency behavior that caused this regression.
- Risk: `pnpm check` is currently red on latest `origin/main` for unrelated TS2883 errors, so full repo green was not available for this PR.
  - Mitigation: scoped verification was rerun after rebase and the unrelated failures were left untouched. Example current failures include `extensions/bluebubbles/src/monitor.webhook.test-helpers.ts`, `src/globals.ts`, and `src/plugin-sdk/provider-auth-login.ts`.

### CODE REVIEW

File: src/docker-setup.e2e.test.ts [hunk 1]
```diff
@@ -206,9 +206,37 @@ describe("scripts/docker/setup.sh", () => {
...
+    const lines = log.split("\n").filter(Boolean);
+    const gatewayStartIdx = lines.findIndex(
+      (line) =>
+        line.includes("compose") && line.includes(" up -d") && line.includes("openclaw-gateway"),
+    );
+    expect(gatewayStartIdx).toBeGreaterThanOrEqual(0);
+
+    const prestartLines = lines.slice(0, gatewayStartIdx);
+    expect(prestartLines.some((line) => line.includes("run --rm openclaw-cli"))).toBe(false);
   });
```
[the remark this answers is not in the document]
Comment (amsminn) [reply src/docker-setup.e2e.test.ts:243]: @greptile-apps I added `await writeFile(activeSandbox.logPath, "")`
before `runDockerSetup(...)` in that test, so the assertion is now
scoped to the current run instead of shared accumulated log state.

Re-ran `corepack pnpm test -- src/docker-setup.e2e.test.ts`; all 20
tests passed.

Fixed in 96b3f78e7d.

File: src/docker-setup.e2e.test.ts [hunk 2]
```diff
@@ -206,9 +206,37 @@ describe("scripts/docker/setup.sh", () => {
...
+
+    const result = runDockerSetup(activeSandbox);
+    expect(result.status).toBe(0);
+
+    const log = await readFile(activeSandbox.logPath, "utf8");
+    const lines = log.split("\n").filter(Boolean);
+    const gatewayStartIdx = lines.findIndex(
+      (line) =>
+        line.includes("compose") && line.includes(" up -d") && line.includes("openclaw-gateway"),
+    );
```
[the remark this answers is not in the document]
Comment (amsminn) [reply src/docker-setup.e2e.test.ts:236]: @chatgpt-codex-connector Fixed in 96b3f78e7d.

The test now clears `activeSandbox.logPath` before invoking
`runDockerSetup(...)`, so the pre-start `openclaw-cli` assertion no
longer depends on log output from earlier tests. I reran `corepack
pnpm test -- src/docker-setup.e2e.test.ts`; 20 tests passed.

File: src/docker-setup.e2e.test.ts [hunk 3]
```diff
@@ -206,9 +206,38 @@ describe("scripts/docker/setup.sh", () => {
...
+    expect(gatewayStartIdx).toBeGreaterThanOrEqual(0);
+
+    const prestartLines = lines.slice(0, gatewayStartIdx);
+    expect(prestartLines.some((line) => line.includes("run --rm openclaw-cli"))).toBe(false);
```
[the remark this answers is not in the document]
Comment (amsminn) [reply src/docker-setup.e2e.test.ts]: I broadened the matcher to fail on any pre-start `docker compose ...
run ... openclaw-cli` invocation before the gateway startup line. I
also reran `corepack pnpm test -- src/docker-setup.e2e.test.ts`; all
20 tests passed.

Fixed in 45cb3ca78d.
